### DISCUSSION TITLE
how to write to AppData\Local\App folder instead of PackageName\AppData\Local\App

### DISCUSSION THREAD

**Discussion by roshannzth** [microsoft/WindowsAppSDK#4018] [Q&A] [answered]
I am having a packaged win ui 3 app whose main purpose is to authenticate user and store idtoken in a cred.json file.
This cred.json file needs to be present in AppData\Local\App instead of PackageName\AppData\Local\App.

The reason being, i have another win32 dll, which loads into windows registration server, and calls few apis with the stored idtoken in cred.json.

Unfortunately, Win ui app creates its own private space in PackageName\AppData\Local\App.
And this dll, finds the AppData\Local\App empty.

So what do i do here to make my win ui app use AppData\Local\App folder.

**Comment by DarranRowe** [comment, marked as answer]
This is a function of the application package itself. This behaviour [is documented](https://learn.microsoft.com/en-us/windows/msix/desktop/desktop-to-uwp-behind-the-scenes#common-file-system-operations). Your only real option is to disable Windows' virtualization of AppData. The steps to do this are [also documented](https://learn.microsoft.com/en-us/windows/msix/desktop/flexible-virtualization).

**Reply by roshannzth** [reply]
@DarranRowe 

The following change to the Package.appxmanifest file did the job
```
<Package
    xmlns:desktop6="http://schemas.microsoft.com/appx/manifest/desktop/windows10/6"
    xmlns:virtualization="http://schemas.microsoft.com/appx/manifest/virtualization/windows10"
    IgnorableNamespaces="rescap desktop6 virtualization">
```
```
    <Properties>
        <desktop6:FileSystemWriteVirtualization>disabled</desktop6:FileSystemWriteVirtualization>

        <virtualization:FileSystemWriteVirtualization>
            <virtualization:ExcludedDirectories>
            <virtualization:ExcludedDirectory>$(KnownFolder:LocalAppData)\App</virtualization:ExcludedDirectory>
            </virtualization:ExcludedDirectories>
        </virtualization:FileSystemWriteVirtualization>
    </Properties>
```
```
    <Capabilities>
        <!-- Include the required restricted capability. -->
        <rescap:Capability Name="unvirtualizedResources"/>
    </Capabilities>
</Package>
```

**Comment by jonwis** [comment]
Maybe you could use CredVault to store the credential token instead of a file on the filesystem? Then you could avoid poking out of hte virtualization bubble.  See https://learn.microsoft.com/windows/win32/api/wincred/nf-wincred-credwritea for how to securely store a credential user-wide with an identifier of your choosing as the `TargetName`.

**Reply by roshannzth** [reply]
@jonwis This can be used for the basic use case of storing creds, but in my case, i had to go with unvirtualized approach just because i do have some bulkdata, like files, logs, etc which is stored and need to be accessed from another binary.

Other approach for storing creds was to use registry, but your approach looks better for the basic use case of storing details.

**Reply by DrusTheAxe** [reply]
> and need to be accessed from another binary

If the other app is running at MediumIL (or higher) it can access your applicationdata via [ApplicationDataManager.CreateForPackageFamily](https://learn.microsoft.com/uwp/api/windows.management.core.applicationdatamanager.createforpackagefamily)(_...pkgfamilyname...).LocalCacheFolder

CredVault's the better answer for sensitive data but for your 'bulk data' this avoids the need to party outside your package in ways Windows won't cleanup if/when you get removed or reset.
